### PR TITLE
[FIX] resource: fix leaves display in gantt

### PR DESCRIPTION
**Pull request by damrOdoo**
current behavior : When an employee has some personnal leaves on a calendar, and the employee changes his calendar, the leaves of the old calendar will not be displayed in the gantt view of the planning.

behavior after the commit : The leaves showed in the gantt view of the planning are now the global leaves of the current calendar of the user + all the personnal leaves of the user, no matter on which calendar those are registered

task 2687135

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/resource/models/resource.py [hunk 1]
```diff
@@ -987,6 +989,78 @@ def _get_unavailable_intervals(self, start, end):
...
+                else:
+                    end = end_dt.astimezone(tz)
+                    tz_dates[(tz, end_dt)] = end
+                dt0 = string_to_datetime(leave.date_from).astimezone(tz)
+                dt1 = string_to_datetime(leave.date_to).astimezone(tz)
+                result[resource.id].append((max(start, dt0), min(end, dt1), leave))
+        mapped_leaves = {r.id: Intervals(result[r.id]) for r in resources_list}
+        return attendances - mapped_leaves[self.id]
+
+    def _get_unavailable_intervals_gantt(self, start_datetime, end_datetime):
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1051]: Isn't it possible to work in batch? :thinking:

File: addons/resource/models/resource.py [hunk 2]
```diff
@@ -17,6 +17,8 @@
 from odoo.tools.float_utils import float_round
 
 from odoo.tools import date_utils, float_utils
+from odoo.tools.safe_eval import pytz
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/resource/models/resource.py:20]: but it is not needed since we already import both `utc` and `timezone` from it. so you can use them directly :)

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Courache :kissing:

**Comment by auon-odoo** [comment]
Also, might be useful to mention https://github.com/odoo/enterprise/pull/15091 in you commit message.

File: addons/resource/models/resource.py [hunk 3]
```diff
@@ -987,6 +989,87 @@ def _get_unavailable_intervals(self, start, end):
             resource_mapping.update(resources_unavailable_intervals)
         return resource_mapping
 
+    def _leaves_interval_gantt_view(self, start_dt, end_dt, leaves, attendances):
```
Comment (tivisse) [inline addons/resource/models/resource.py:992]: If this code is related to gantt view, I don't get why it is defined in a low-level community application.

Comment (damrOdoo) [reply addons/resource/models/resource.py:992]: I did not find a better module to do so. Since this change affect project_enterprise and planning, i had to find a common module for both of them to avoid duplicating the code. The module resource seemed more appropriate than the web_gantt, which seems to be very generic too.

File: addons/resource/models/resource.py [hunk 4]
```diff
@@ -987,6 +988,82 @@ def _get_unavailable_intervals(self, start, end):
...
+
+        start_dt = utc.localize(start_datetime) if not start_datetime.tzinfo else start_datetime
+        end_dt = utc.localize(end_datetime) if not end_datetime.tzinfo else end_datetime
+        # both search are made in batch. probably a good idea to mapped the personnal leaves per resource_id in order to not iterate over all leaves every time.
+        personal_leaves_intervals = self._personal_leaves_intervals(start_dt, end_dt)
+        global_leave_per_calendar = self._global_leaves_per_calendar(start_dt, end_dt, calendars_set)
+        dict_attendance_per_calendar = {}
+        for calendar in calendars_set:
+            attendances = calendar._attendance_intervals_batch(start_dt, end_dt, resources=self)
+            dict_attendance_per_calendar[calendar] = attendances
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1025]:
```suggestion
        attendance_per_calendar = {
           calendar: calendar._attendance_intervals_batch(start_dt, end_dt, resources=self)
           for calendar in res.calendar_id
        }
```
Et pouf!

File: addons/resource/models/resource.py [hunk 5]
```diff
@@ -987,6 +988,82 @@ def _get_unavailable_intervals(self, start, end):
...
+
+    def _personal_leaves_intervals(self, start_dt, end_dt):
+        resource_ids = [r.id for r in self]
+        domain = [
+            ('time_type', '=', 'leave'),
+            ('resource_id', 'in', resource_ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        return self.env['resource.calendar.leaves'].search(domain)
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1048]:
```suggestion
        return self.env['resource.calendar.leaves'].search([
            ('time_type', '=', 'leave'),
            ('resource_id', 'in', self.ids),
            ('date_from', '<=', end_dt),
            ('date_to', '>=', start_dt),
        ])
```
1) Create variable if you use the same value multiple times or as a way to fragment/document your code when it is getting complex. In this case, it really isn't worth it.
2) `[r.id for r in self]` = `self.ids` :scream:

File: addons/resource/models/resource.py [hunk 6]
```diff
@@ -987,6 +988,82 @@ def _get_unavailable_intervals(self, start, end):
...
+                ('time_type', '=', 'leave'),
+                ('calendar_id', 'in', [False, calendar_id]),
+                ('date_from', '<=', end_dt),
+                ('date_to', '>=', start_dt),
+            ]
+
+        for calendar in calendars_set:
+            leaves = self.env['resource.calendar.leaves'].search(_get_domain(calendar.id))
+            dict_global_leaves[calendar] = leaves
+        return dict_global_leaves
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1066]:
```suggestion
    def _global_leaves_per_calendar(self, start_dt, end_dt, calendars_set):
        Leave = self.env['resource.calendar.leaves']
        Calendar = self.env['resource.calendar']
        return {
            Calendar.browse(group['calendar_id']): Leave.browse(group['ids'])
            for group in Leave._read_group(
                domain=[
                    ('resource_id', '=', False),
                    ('time_type', '=', 'leave'),
                    ('calendar_id', 'in', [False] + calendars_set.ids),
                    ('date_from', '<=', end_dt),
                    ('date_to', '>=', start_dt),
                ],
                fields=['ids:array_agg(id)'],
                groupby=['calendar_id']
            )
        }
```
Doesn't that work? So that we don't make a request at each iteration.

File: addons/resource/tests/test_resource.py [hunk 7]
```diff
@@ -1220,3 +1221,46 @@ def test_unavailable_intervals(self):
...
+    def test_unavailable_intervals_gantt(self):
+        """
+            this test ensure that when an employee change from one calendar to another, all the personnal time-off taken from the
+            previous calendar are correctly displayed on the gantt view.
+            1. Set a time-off in calendar A for employee A.
+            2. Set a global leave on calendar A
+            3. Set a global leave on calendar B.
+            4. Check that the return value contains the time-off of employee A from calendar A, the global time-off from
+            calendar B, and not the global time-off from calendar A.
+        """
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/resource/tests/test_resource.py:1234]: Don't you agree? :sweat_smile:

File: addons/resource/models/resource.py [hunk 8]
```diff
@@ -17,6 +17,7 @@
 from odoo.tools.float_utils import float_round
 
 from odoo.tools import date_utils, float_utils
+
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:20]:
```suggestion
```
Unnecessary change

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Getting better... :fire:

File: addons/resource/models/resource.py [hunk 9]
```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
...
+            else:
+                start = start_dt.astimezone(tz)
+                tz_dates[(tz, start_dt)] = start
+            if (tz, end_dt) in tz_dates:
+                end = tz_dates[(tz, end_dt)]
+            else:
+                end = end_dt.astimezone(tz)
+                tz_dates[(tz, end_dt)] = end
+
+            personal_leaves = personal_leaves_intervals[res.id] if personal_leaves_intervals.get(res.id) else False
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1050]:
```suggestion
            personal_leaves = personal_leaves_intervals.get(res.id, False)
```
2nd arg in get is the default/"or_else" value. :slightly_smiling_face:

File: addons/resource/models/resource.py [hunk 10]
```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
...
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'resource_id'])
+        dict_personal_leaves = {}
+        for leave in leaves:
+            if dict_personal_leaves.get(leave['resource_id'][0]):
+                dict_personal_leaves[leave['resource_id'][0]].append([leave['date_from'], leave['date_to']])
+            else:
+                dict_personal_leaves[leave['resource_id'][0]] = [[leave['date_from'], leave['date_to']]]
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1080]:
```suggestion
        leaves_per_resource = defaultdict(list)
        for leave in leaves:
            dict_personal_leaves[leave['resource_id'][0]].append((leave['date_from'], leave['date_to']))
```
- The default kinda works the same way as `get` (see above). Its arg is the default/"or_else" value it will set on the key that you try to change but isn't yet in the dict.
- Make tuples if you don't plan to update them. :ok_hand: 
- Also, name dict \<value\>_per\<key\> please.

File: addons/resource/models/resource.py [hunk 11]
```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
...
+        no calendar_id
+        """
+        domain = [
+            ('resource_id', '=', False),
+            ('time_type', '=', 'leave'),
+            ('calendar_id', 'in', [False] + calendars_set.ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'calendar_id'])
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1095]: Can't we group the request with the previous one? Records will be easy to distinguish base on their resource_id. :thinking:

File: addons/resource/models/resource.py [hunk 12]
```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
...
+            key = leave['calendar_id'][0] if leave['calendar_id'] else False # the 'if' is needed in case of no calendar_id. It avoids a traceback on 'leave['calendar_id'][0]'
+            if dict_global_leaves.get(key):
+                dict_global_leaves[key].append([leave['date_from'], leave['date_to']])
+            else:
+                dict_global_leaves[key] = [[leave['date_from'], leave['date_to']]]
+        # in case some global leaves are not linked to any calendar, we must add them to each dict entry to ensures they are computed for each user
+        if not dict_global_leaves.get(False):
+            return dict_global_leaves
+        common_leaves = dict_global_leaves[False]
+        for key in dict_global_leaves:
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:1107]: And this could be done in the previous loop, right? :slightly_smiling_face:

File: addons/resource/models/resource.py [hunk 13]
```diff
@@ -1,13 +1,12 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
-
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:3]: Woopsy, unnecessary change...

File: addons/resource/models/resource.py [hunk 14]
```diff
@@ -1,13 +1,12 @@
...
 from collections import defaultdict
 import math
 from datetime import datetime, time, timedelta
 from dateutil.relativedelta import relativedelta
 from dateutil.rrule import rrule, DAILY, WEEKLY
 from functools import partial
-from itertools import chain
+from itertools import chain, groupby
```
Comment (auon-odoo) [inline addons/resource/models/resource.py:9]: This import in not used :angel:

**Review by auon-odoo — CHANGES_REQUESTED** [review]
It is getting better. Love it!
Thanks for your work, keep it up :smile:

File: addons/resource/models/resource.py (hunk 12 above)
Comment (damrOdoo) [reply addons/resource/models/resource.py:1107]: I dont think it's doable. Say for example, you have 2 leave in the recordset, and you iterate through it. The first is a leave with no calendar_id, the second is  a leave with calendar_id = 2. 
The iteration will first create an entry for the key 'False', but it wont be possible to add it to the key  '2' since it has not been created yet.

File: addons/resource/models/resource.py (hunk 11 above)
Comment (damrOdoo) [reply addons/resource/models/resource.py:1095]: No that's the core of the problem. If we take the following case : one leave A in calendar B, and the user U with calendar C.
We need both the leave A and the global leave of calendar C. If we merge the two domain, we will excludes the calendar B from the search.
If you really want to merge the domain and do only one search, we'll have to fetch all the leave of the db with something like 
```
domain = [
            ('time_type', '=', 'leave'),
            ('date_from', '<=', end_dt),
            ('date_to', '>=', start_dt),
        ]
```
which does not seems like a good idea.

Comment (auon-odoo) [reply addons/resource/models/resource.py:1095]: Clarified in mp

File: addons/resource/models/resource.py (hunk 12 above)
Comment (auon-odoo) [reply addons/resource/models/resource.py:1107]: Nvm, you are right. :see_no_evil:

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Last changes :kissing_heart:

**Review by auon-odoo — APPROVED** [review]
Lgtm. Many thanks :fire: :fire:

**Comment by C3POdoo** [comment]
Dear @damrOdoo,

Thank you for your contribution but the version 14.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._
